src/olympia/addons/serializers.py

```diff
@@ -812,7 +850,26 @@ def get_is_source_public(self, obj):
 
     def validate(self, data):
         if not self.instance:
-            addon_type = self.fields['version'].parsed_data['type']
+            parsed_data = self.fields['version'].parsed_data
+            addon_type = parsed_data['type']
+            channel = getattr(data.get('version', {}).get('upload'), 'channel', None)
+
+            # If this is a new addon, check that all the required metadata is set.
+            # We test for new versions in VersionSerailizer.validate instead
+            if channel == amo.RELEASE_CHANNEL_LISTED:
+                # This is replicating what Addon.get_required_metadata does
+                required_msg = 'This field is required for addons with listed versions.'
+                missing_metadata = {
```

Comment:
This is for new add-ons with a listed version, where we check that the metadata was either in the POST (`data`) or in the manifest (`parsed_data`) to make the add-on complete.  

For new versions on an existing add-on (e.g. a listed version on an add-on that was previously submitted with an unlisted version) there isn't any update to the add-on, so this code isn't called.

Reply:
re: `field`: `missing_metadata` is set using a a dict generator, so `field` is the key, and `required_msg` will be the value, so if, for example, `name` and `summary` aren't provided in the manifest or `POST`, it will generate a dict like:
```python
{
    'name': 'This field is required for addons with listed versions.',
    'summary': 'This field is required for addons with listed versions.',
}
```
`ValidationError` can accept a dict as an arg which outputs the validation errors correctly for multiple fields.

Reply:
> why this code is different

so, regarding the subtle difference, in `VersionSerializer` we already have the add-on so can check if the metadata is already provided, but we can't raise `ValidationError` on each missing field because that data wasn't provided with the request - if you're submitting a new version you're only concerned with the version, e.g. you can't set the add-on's categories at the same time - so we have to error with a message saying those values need to be set first via the addon api